README.md
# The Attacker IP Prioritizer V2.0.3

## The Idea

The Attacker IP Prioritizer (AIP) algorithm aims to generate a IoT friendly blacklist. With the advent of 5G, many IoT devices are going to be directly connected to the internet instead of being protected by a routers firewall. Therefore we need blacklists that are small and portable, and designed to blacklist IPs that are targeting IoT. Blacklists are not the be all end all of security since they are past-data bound, and as such they can never fully protect a device, and a full IPS solution is usually needed. However, if Blacklists contain the IPs that are the most aggressive and dangerous, most attacks can be stopped, thus lowering the stress on the full IPS solution. AIP aims at providing a data and statistics driven blacklist that only contains the IPs that NEED to be blocked.

## Data Source

The format for the input data is a directory that contains data files from each day. You assign a directory for the program to look in every time it runs, and it checks if there are any new files to process. If there are, it processes the new files and remembers the names of the new files so that it does not process it the next time it runs.

In terms of file format, it accepts a .csv file that has one IP per line, with each of the following data inputs for each IP on that line, separated by commas:

    Amount of events - Meaning the total connections to our honeypots originating from the given IP

    Total Duration - How long did this IP connect for the total of its events

    Average duration - The average length in seconds of all the connections per IP

    Amount of Bytes - Total bytes sent and received

    Average number of bytes - For bytes transferred in each connection per IP

    Total packets - Of all the connections per IP

    Average packets - Average packets sent per connection

    Last event time - UNIX time of the last time the IP tried to connect to something  in the last 24 hours

    First event time - UNIX time of the first time the IP tried to connect in the last 24 hours

For example, a single line in the file could look like this:

"IPv4 Address",26049,"7415310","284.6","41808957","1605.0",284577,"10.92","157899154","1578968762.519"


ChangeLog.md
16/10/2020
V2.0.3
- Updated Whitelist for master branch
- Updated README and AIP-How-To-Guide

07/10/2020
V2.0.2
- Created branch V2.0.2 for testing
- Fixed problem with parenthesis in the blacklist headers
- Merged V2.0.2 with master

18/09/2020
- Modified thresholds to shrink lists

17/09/2020
- Moved the daily blacklists to a separate repository

15/09/2020
- Created branch V3.0.0 for next major update

24/07/2020
V2.0.1
- Minor bug fixes

22/07/2020
V2.0.0
- Added a whitlist module, namely before using the main data file, program will check for IPs that should not be blacklisted
- Added an auto-run script.
- Changed the aging function so it keeps track og how much an IP needs to be reduced over time
- Changed the threshold so that it is dynamic again
- Fixed the cannot check whitelist bug
- Fixed the infinite loop bug for the prioritize new normalized rating function
- Fixed other small bugs and improved code

07/05/2020
V1.0.3
- Updated README.md files
- Updated AIP-How-To-Guide.md

28/04/2020
V1.0.2
- Fixed problem with calling modules incorrectly
- Add ability to choose new instance, or running instance
- Asks for input data file location
- Checks if input directory exits
- If directoy exist, creates needed folders in it, if it does not, asks if we wisd to continue, and creates it all
- If not a new instance, simple asks where the old instance is.
- IMPORTANT - There is a bug with updating an old instance!

24/04/2020
V1.0.1
- Separated eval module into another repo
- Separated eval script and run AIP script
- Worked on organization
- Deleted unecessary functions in AIP.py
- Removed old module files

02/04/2020
V1.0.0
- First major commit, via the correct methods
- Removed old files that were no longer relevant
- Committed Main and Run-AIP.sh, but not Eval and Total_Eval.sh because both still contain personal identifiers


generate_absolute_data.sh
#!/bin/bash
#!/bin/bash
answer1="Yes"
answer2="No"
answer3="1"
answer4="2"

echo "............................"
echo "Please Select a option from the menu[1,2]:"
echo "(1) New Instance, create ouput folder and folders"
echo "(2) Running Instance, save output to an existing folder"
read option
echo "............................"

echo "Location of input data files(Can contain one or more files):"
read input_data_folder
echo "............................"

while [ ! -d $input_data_folder ]
do
   echo Not a real directory, please input a real one:
   read input_data_folder
   echo ............................
done

# If we are going to create a new instance, we need to create all the needed record files
if [ "$option" = "$answer3" ]
then
		echo New Directory where you want results, no slash at end:
		read output_folder
		echo ............................
		# Check to see if the directory is real.
		while [ ! -d $output_folder ]
		do
			echo Not a real directory, do you wish to create it? [Yes,No]
			read answer
		  if [ "$answer" = "$answer1" ]
		  then
		  	mkdir $output_folder
				mkdir $output_folder/Input_Data/
				mkdir $output_folder/Updated_Data_By_Day/
				touch $output_folder/Absolute_Data.csv
				touch $output_folder/Known_IPs.txt
				touch $output_folder/Processed_Splunk_Files.txt
				touch $output_folder/Times.csv
		  elif [ "$answer" = "$answer2" ]
		  then
		  	echo Input different location:
		  	read output_folder
				mkdir $output_folder/Input_Data/
				mkdir $output_folder/Updated_Data_By_Day/
				touch $output_folder/Absolute_Data.csv
				touch $output_folder/Known_IPs.txt
				touch $output_folder/Processed_Splunk_Files.txt
				touch $output_folder/Times.csv
		  else
		  	continue
			echo ............................
		  fi
		done
elif [ "$option" = "$answer4" ]
then
	echo "Please Input location of current instance ouput files:"
	read output_folder
fi

# Export all variables so they can be accessed by AIP.py
export output_folder
export input_data_folder

directory_of_AIP=$(dirname $(readlink -f "generate_absolute_data.sh"))

echo $directory_of_AIP

for entry in $input_data_folder/*
do
   cp "$entry" $output_folder/Input_Data/
   echo "$entry"
   python3 $directory_of_AIP/Main/update_historical_data.py
   cp $output_folder/Absolute_Data.csv $output_folder/Updated_Data_By_Day/`ls -1 $entry | grep -oP '[\d]+-[\d]+-[\d]+'`_data_file.csv
done


Main/update_historical_data.py
import csv
from datetime import datetime
from whitelist_module import load_whitelist, check_if_ip_is_in_whitelisted_nets, check_if_ip_is_in_whitelisted_ips
import os

startTime = datetime.now()

# Full path to directory where all the files will be stored
# (a)
AIPP_direcory = os.environ['output_folder']

# Full path to the  folder where the program will look for new data files. It will look in the file and only process the
# files it has not precessed yet. It will process every file it does not recognize.
# (b)
folder_path_for_raw_Splunk_data = AIPP_direcory + '/Input_Data'

# Full path to the file where the program will record the data files it processes
# (c)
record_file_path_for_processed_Splunk_files = AIPP_direcory + '/Processed_Splunk_Files.txt'

# A complete list of every IP seen by the program since it was started
# (d)
record_file_path_to_known_IPs = AIPP_direcory + '/Known_IPs.txt'

# Full path to the file where the data flows for each IP are stored. Includes all the data the program has received
# since it was started. This is NOT the file that contains the ratings.
# (e)
record_file_path_for_absolute_data = AIPP_direcory + '/Absolute_Data.csv'

# Full path to folder that wil contain the daily rating files. This is a FOLDER!!
# (f)
directory_path_historical_ratings = AIPP_direcory + '/Historical_Ratings'

def find_new_data_files(b, c):
    list_of_new_data_files = []
    list_of_data_files = os.listdir(b)
    dictionary_of_dates_on_files = {}
    with open(c, 'r') as record:
        list_of_processed_data_files = record.read().split('\n')
    for file in list_of_data_files:
        if file not in list_of_processed_data_files:
            list_of_new_data_files.extend([file])
            dictionary_of_dates_on_files[file[0:10]] = file
    for file12 in list_of_new_data_files:
        with open(c, 'a') as records_file1:
            records_file1.write(file12 + '\n')
    sorted_dates = sorted(dictionary_of_dates_on_files, key=lambda date: datetime.strptime(date, '%Y-%m-%d'))
    sorted_dates.reverse()
    print(sorted_dates)
    return list_of_new_data_files, sorted_dates[0]

# >>>>>>>>>>>>>>> Call the find new file function and define the time reference point for the aging function
new_data_files, date = find_new_data_files(folder_path_for_raw_Splunk_data, record_file_path_for_processed_Splunk_files)
print('There are ', len(new_data_files), ' new data files to process')
current_time = datetime(int(date[0:4]), int(date[5:7]), int(date[8:10]), 1).timestamp()

# File that will be storing the run times for this script
time_file = AIPP_direcory + '/Times.csv'

def open_sort_new_file(b, list_of_new_files):
    list_of_new_data_flows = []
    list_of_IPs_in_new_data = []
    for file in list_of_new_files:
        with open(b + '/' + file, 'r') as csvfile:
            for line in csv.reader(csvfile):
                if line[0] != 'SrcAddr':
                    list_of_new_data_flows.append([line[0], line[1], line[2], line[3], line[4], line[5], line[6], line[7],
                                                   line[8], line[9], line[10], line[11], line[12], line[13], line[14], line[15], line[16]])
                    list_of_IPs_in_new_data.append(line[0])
                else:
                    continue
    return list_of_new_data_flows, list_of_IPs_in_new_data


def open_sort_abs_file(e):
    IP_flows = []
    IPs_in_absolute_file = []
    with open(e, 'r') as csvfile:
        for line in csv.reader(csvfile):
            if not line:
                break
            else:
                IP_flows.append([line[0], line[1], line[2], line[3], line[4], line[5], line[6], line[7], line[8], line[9], line[10]])
                IPs_in_absolute_file.append(line[0])
    return IP_flows, IPs_in_absolute_file

def get_updated_flows(location_of_absolute_data_file):
    IP_flows = []
    with open(location_of_absolute_data_file, 'r') as csvfile:
        for line in csv.reader(csvfile):
            if not line:
                break
            else:
                IP_flows.append([line[0], line[1], line[2], line[3], line[4], line[5], line[6], line[7], line[8], line[9], line[10]])
    return IP_flows

def sort_IPs_from_data(IPs_from_absolute_data, IP_flows_from_todays_data):
    unknown_IP_flows = []
    unknown_IPs = []
    known_IPs = []
    known_IP_data_flows = []
    for IP in IP_flows_from_todays_data:
        if IP[0] in IPs_from_absolute_data:
            known_IP_data_flows.append(IP)
            known_IPs.append(IP[0])
        elif IP[0] not in IPs_from_absolute_data:
            unknown_IP_flows.append(IP)
            unknown_IPs.append(IP[0])
    return unknown_IP_flows, unknown_IPs, known_IP_data_flows, known_IPs


def write_unkown_IPs_to_data_file(list_of_unknown_IPs, d):
    with open(d, 'a') as data_file:
        for flow in list_of_unknown_IPs:
            data_file.write(flow + '\n')


def update_records_files(e, list_of_known_new_IP_data, unknown_IP_flows):
    absolute_data, IPs_in_abs_file = open_sort_abs_file(e)
    new_absolute_file_flows = []
    new_absolute_file_flows += absolute_data
    unknown_IP_flows_new = []
    for new_flow2 in unknown_IP_flows:
        new_flow2.extend([new_flow2[1]])
        unknown_IP_flows_new.append(new_flow2)
    new_absolute_file_flows.extend(unknown_IP_flows_new)

    if not list_of_known_new_IP_data:
        pass

    else:
        for x1, new_flow in enumerate(list_of_known_new_IP_data):
            for x2, absolute_flow in enumerate(new_absolute_file_flows):
                if absolute_flow[0] == new_flow[0]:
                    days_since_first_seen = (current_time - float(absolute_flow[9])) // 86400.0
                    if days_since_first_seen != 0:
                        updated_event_average = ((float(absolute_flow[10])) * (days_since_first_seen - 1) + float(
                            new_flow[1])) / days_since_first_seen
                    else:
                        updated_event_average = ((float(absolute_flow[10])) * (days_since_first_seen - 1) + float(
                            new_flow[1])) / 1

                    updated_total_events = float(new_flow[1]) + float(absolute_flow[1])
                    updated_total_duration = float(absolute_flow[2]) + float(new_flow[2])
                    updated_average_duration = (float(absolute_flow[3]) + float(new_flow[3])) / 2.0
                    updated_total_bytes = float(absolute_flow[4]) + float(new_flow[4])
                    updated_average_bytes = (float(absolute_flow[5]) + float(new_flow[5])) / 2.0
                    updated_total_packets = float(absolute_flow[6]) + float(new_flow[6])
                    updated_average_packets = (float(absolute_flow[7]) + float(new_flow[7])) / 2.0
                    updated_last_event = new_flow[9]

                    updated_entry = [new_flow[0], updated_total_events, updated_total_duration,
                                     updated_average_duration,
                                     updated_total_bytes, updated_average_bytes, updated_total_packets,
                                     updated_average_packets,
                                     absolute_flow[8], updated_last_event, updated_event_average]
                    new_absolute_file_flows[x2] = updated_entry
                    break
                else:
                    continue

    whitelisted_nets, whitelisted_ips = load_whitelist()
    for index, flow in enumerate(new_absolute_file_flows):
        judgement1 = check_if_ip_is_in_whitelisted_nets(flow[0], whitelisted_nets)
        judgement2 = check_if_ip_is_in_whitelisted_ips(flow[0], whitelisted_ips)
        if (judgement1==True) or (judgement2==True) is True:
            del new_absolute_file_flows[index]
            print('Found ', flow[0], ' in Whitelisted IPs. Deleting entry...')
        else:
            continue

    with open(e, 'w') as new_file_another:
        wr2 = csv.writer(new_file_another, quoting=csv.QUOTE_ALL)
        for y in new_absolute_file_flows:
            wr2.writerow(y)

# Now call all the functions on the data

list_of_known_data_flows, list_of_known_IPs_in_data = open_sort_abs_file(record_file_path_for_absolute_data)

list_of_new_data_flows, list_of_IPs_in_new_data = open_sort_new_file(folder_path_for_raw_Splunk_data, new_data_files)

unknown_IP_flows_from_new_data, unknown_IPs_from_new_data, known_IP_data_flows_from_new_data, known_IPs_from_new_data\
    = sort_IPs_from_data(list_of_known_IPs_in_data, list_of_new_data_flows)

write_unkown_IPs_to_data_file(unknown_IPs_from_new_data, record_file_path_to_known_IPs)

update_records_files(record_file_path_for_absolute_data, known_IP_data_flows_from_new_data, unknown_IP_flows_from_new_data)

# Append the time that it took to a file
with open(time_file, 'a') as new_file_another:
        wr2 = csv.writer(new_file_another, quoting=csv.QUOTE_ALL)
        list4 = []
        list4.append(date)
        list4.append(datetime.now() - startTime)
        wr2.writerow(list4)